### PR TITLE
feat: Support Arrow PyCapsule Interface for export

### LINKED ISSUE #784: [Enh]: Pass through the Arrow PyCapsule Interface
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

The [Arrow PyCapsule Protocol](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html) makes it possible to share Arrow data between libraries without requiring a dependency on pyarrow. Many of the DataFrame libraries that narwhals wraps already support it (See https://github.com/apache/arrow/issues/39195 for ecosystem summary), so my thinking is that it wouldn't be too much work to support the protocol on the Narwhals DataFrame itself.

### Please describe the purpose of the new feature or describe the problem to solve.

Vega-Altair recently adopted Narwhals as a way to support polars without a pyarrow dependency, and I'm looking at how I could do the same in VegaFusion. The next VegaFusion release is going to support ingesting DataFrames that support the PyCapsule Interface (WIP by @kylebarron in https://github.com/vega/vegafusion/pull/501), so it would be nice to be able to pass the Narwhals DataFrame directly to VegaFusion and process it using this protocol.

### Suggest a solution if possible.

_No response_

### If you have tried alternatives, please describe them below.

_No response_

### Additional information that may help us understand your needs.

_No response_

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #784 

@kylebarron fancy taking a look to see if this is what needs doing / if I've understood the assignment?

## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Review by kylebarron — APPROVED** [review]
Yeah I think that makes sense! There's not much you can do if the underlying dataframe doesn't support it. If you had some way to access an arrow table from the underlying dataframe you could do something more, but I think this is good

**Comment by MarcoGorelli** [comment]
cool, thanks!

> If you had some way to access an arrow table from the underlying dataframe

we do have `narwhals.DataFrame.to_arrow`, which returns a pyarrow table - is that what you meant? If so, what else would you suggest adding?

**Comment by kylebarron** [comment]
In that case I'd suggest
```py
def __arrow_c_stream__(self, requested_schema: object | None = None) -> object:
    try:
        return self._compliant_frame._native_frame.__arrow_c_stream__(
            requested_schema=requested_schema
        )
    except AttributeError:
        pyarrow_table = self.to_arrow()
        return pyarrow_table.__arrow_c_stream__(requested_schema=requested_schema)
```
So in the first case, we can use the source's implementation of the pycapsule interface and pyarrow doesn't need to be installed, while in the second case we can still ensure the method never raises an exception.

**Comment by MarcoGorelli** [comment]
thanks! though I think the user could call this themselves? e.g.
```python
df: nw.DataFrame
try:
    result = df.__arrow_c_stream__(requested_schema=requested_schema)
except AttributeError:
    result = df.to_arrow().__arrow_c_stream__(requested_schema=requested_schema)
```
?
I think that'd be more explicit, not totally sure we should be calling `to_arrow` on behalf of the user. By 'user', I mean the library developer using Narwhals, I'd be inclined to leave it up to them whether or not to fallback to PyArrow

**Comment by kylebarron** [comment]
Well, one of my primary arguments for the pycapsule interface is that it allows an ecosystem of data producers and consumers to interoperate without any knowledge of the other, solely by looking for an `__arrow_c_stream__` dunder method. Calling `.to_arrow()` would indeed be more explicit, but it would require library consumers to _know about narwhals_, which I'd argue in the general case is not true. E.g. `pyarrow.table()` only knows to check for `__arrow_c_stream__`.

**Comment by MarcoGorelli** [comment]
I was thinking more of the Vegafusion case - I think it's better for them if they choose to explicitly call `to_arrow`. Otherwise they might be calling `__arrow_c_stream__` all over the places, thinking it's cheap, whereas it would've been better to do a single `to_arrow` upfront 😇 

I think any library developer using Narwhals (e.g. vegafusion) would and should know about Narwhals, whereas lower-level libraries like PyArrow shouldn't:
- PyArrow can just check for `__arrow_c_stream__` (as it currently does)
- Vegafusion can choose whether and when to call `to_arrow` before accessing `__arrow_c_stream__`

**Comment by MarcoGorelli** [comment]
In any case, as they often say, "in open source, 'no' is temporary but 'yes' is forever" - doubly so with our stable api policy 😆 

So, as the current implementation looks good to you, I'd say - let's start with that, we can always loosen it later if necessary

Thanks for your review and input, much appreciated 🙏 !

**Comment by kylebarron** [comment]
> Otherwise they might be calling `__arrow_c_stream__` all over the places, thinking it's cheap, whereas it would've been better to do a single `to_arrow` upfront

As a general note, consumers can't do this because consumers don't have a way to know whether the source is a _table_ that already exists in memory or whether it's a _stream_ that can only be called once. E.g. a pyarrow `RecordBatchReader` is a stream and you can only call `__arrow_c_stream__` once. So for a consumer like vegafusion, it would be important for it to import all the data once and then operate as it needs to on it.

**Comment by MarcoGorelli** [comment]
Ah nice, thanks for explaining!

In that case, I'm leaning more towards your suggestion - it would also mean being able to support this for versions of pandas prior to 2.2 but which support converting to pyarrow table

File: narwhals/series.py [hunk 1]
```diff
@@ -71,7 +71,14 @@ def __arrow_c_stream__(self, requested_schema: object | None = None) -> object:
...
-        pa = get_pyarrow()
+        try:
+            import pyarrow as pa
+        except ModuleNotFoundError as exc:  # pragma: no cover
+            msg = f"PyArrow>=14.0.0 is required for `__arrow_c_stream__` for object of type {type(native_series)}"
+            raise ModuleNotFoundError(msg) from exc
+        if parse_version(pa.__version__) < (14, 0):  # pragma: no cover
+            msg = f"PyArrow>=14.0.0 is required for `__arrow_c_stream__` for object of type {type(native_series)}"
+            raise ModuleNotFoundError(msg)
         ca = pa.chunked_array([self.to_arrow()])
```
Comment (kylebarron) [inline narwhals/series.py:82]: I think this might require pyarrow 15

Comment (MarcoGorelli) [reply narwhals/series.py:84]: in pandas the requirement is PyArrow 14+ (I also just ran the tests with pyarrow 13 and 14 - the former fails, the latter passes)

Comment (MarcoGorelli) [reply narwhals/series.py:84]: ah sorry, that's for DataFrame

Comment (kylebarron) [reply narwhals/series.py:84]: It was added to pa.chunked_array in a later release, yes. I think it was here: https://github.com/apache/arrow/pull/40818

**Comment by MarcoGorelli** [comment]
thanks Kyle for your help!

cool, let's ship this 🚢

**Pull request merged**